Tests/NIOSSLTests/ServerSNITests.swift

```diff
@@ -0,0 +1,154 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import XCTest
+import NIO
+import NIOTLS
+import NIOSSL
+
+
+class ServerSNITests: XCTestCase {
+    
+    class SniServerContext {
+    
+        private var _tlsConfig : TLSConfiguration?;
+        
+        private var _ctx : NIOSSLContext?
+        
+        private(set) var SniExpectation : XCTestExpectation
+        
+        public var SSLContext : NIOSSLContext
+        {
+            get
+            {
+                return _ctx!
+            }
+        }
+        
+        private var _indicatedHostname : String?
+        
+        public var IndicatedHostname : String
+        {
+            get
+            {
+                return _indicatedHostname!
+            }
+        }
+        
+        private init(sniExpectation: XCTestExpectation) {
+            SniExpectation = sniExpectation
+        }
+     
+        private func sniCallback (host: String, ctx: NIOSSLContext) -> NIOSSLContext? {
+            self._indicatedHostname = host
+            self.SniExpectation.fulfill()
+            return ctx
+        }
+        
+        public static func Create(sniExpectation: XCTestExpectation) -> SniServerContext {
+            let serverCtx = SniServerContext(sniExpectation: sniExpectation)
+            
+            serverCtx._indicatedHostname = String()
+            serverCtx._tlsConfig = TLSConfiguration.makeServerConfiguration(sniCallback: serverCtx.sniCallback)
+            serverCtx._ctx = try! NIOSSLContext(configuration: serverCtx._tlsConfig!)
+            
+            return serverCtx
+        }
+    }
+    
+    static var cert: NIOSSLCertificate!
+    static var key: NIOSSLPrivateKey!
+
+    override class func setUp() {
+        super.setUp()
+        let (cert, key) = generateSelfSignedCert()
+        NIOSSLIntegrationTest.cert = cert
+        NIOSSLIntegrationTest.key = key
+    }
+
+    private func configuredClientSSLContext() throws -> NIOSSLContext {
+        var config = TLSConfiguration.makeServerConfiguration(
+            certificateChain: [.certificate(NIOSSLIntegrationTest.cert)],
+            privateKey: .privateKey(NIOSSLIntegrationTest.key)
+        )
+        config.trustRoots = .certificates([NIOSSLIntegrationTest.cert])
+        let context = try NIOSSLContext(configuration: config)
+        return context
+    }
+    
+    private func configuredServerSSLContext() throws -> SniServerContext {
+        let expectation = expectation(description: "Notification Raised")
+        return SniServerContext.Create(sniExpectation: expectation)
+    }
+
+    private func assertSniResult(sniField: String?, expectedResult: String) throws {
+        let clientContext = try configuredClientSSLContext()
+
+        let serverContext = try configuredServerSSLContext()
+        
+        let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
```

Comment:
Can we try to run these using `EmbeddedChannel` and the `connectInMemory`/`interactInMemory` functions defined in other test helper files? This reduces FD pressure in CI machines and also makes the tests vastly more reproducible, as well as faster.